mesonbuild/mformat.py

```diff
@@ -0,0 +1,966 @@
+            for f in fields(config):
+                if getattr(cp, f.name, None) is not None:
+                    continue  # value already set from higher file
+
+                getter = f.metadata['getter']
+                for section in sections:
+                    value = getter(cp, section, f.name, fallback=None)
+                    if value is not None:
+                        setattr(config, f.name, value)
+
+            if cp.getboolean(cp.default_section, 'root'):
+                break
+
+        return config
+
+    def load_configuration(self, configuration_file: T.Optional[Path]) -> FormatterConfig:
+        config = FormatterConfig()
+        if configuration_file:
+            cp = DefaultConfigParser()
+            cp.read_default(configuration_file)
+
+            for f in fields(config):
+                getter = f.metadata['getter']
+                value = getter(cp, cp.default_section, f.name, fallback=None)
+                if value is not None:
+                    setattr(config, f.name, value)
+
+        return config
+
+    def format(self, code: str, source_file: Path) -> str:
+        self.current_dir = source_file.parent
+        self.current_config = FormatterConfig.default()
+        if self.use_editor_config:
+            self.current_config = self.current_config.with_editorconfig(self.load_editor_config(source_file))
+        self.current_config = self.current_config.update(self.config)
+
+        ast = mparser.Parser(code, source_file.as_posix()).parse()
+        if self.fetch_subdirs:
+            subdir_fetcher = SubdirFetcher(self.current_dir)
+            ast.accept(subdir_fetcher)
+            self.subdirs = subdir_fetcher.subdirs
+
+        ast.accept(AstConditionLevel())
+        for level in range(5):
+            ast.accept(TrimWhitespaces(self.current_config))
+            ast.accept(ArgumentFormatter(self.current_config))
+
+            cll = ComputeLineLengths(self.current_config, level)
+            ast.accept(cll)
+            if not cll.need_regenerate:
+                break
+
+        printer = RawPrinter()
+        ast.accept(printer)
+        return printer.result
+
+
+def add_arguments(parser: argparse.ArgumentParser) -> None:
+    inplace_group = parser.add_mutually_exclusive_group()
+    inplace_group.add_argument(
+        '-q', '--check-only',
+        action='store_true',
+        help='exit with 1 if files would be modified by meson format'
+    )
+    inplace_group.add_argument(
+        '-i', '--inplace',
+        action='store_true',
+        help='format files in-place'
+    )
+    parser.add_argument(
+        '-r', '--recursive',
+        action='store_true',
+        help='recurse subdirs (requires --check-only or --inplace option)',
+    )
+    parser.add_argument(
+        '-c', '--configuration',
+        metavar='meson_fmt.ini',
+        type=Path,
+        help='read configuration from meson_fmt.ini'
```

Comment:
Can we call this `meson.format` instead of `meson_fmt.in`?

Reply:
Done. I also added the feature to automatically load `meson.format` when it is present and no config file is specified on the command line.